### ISSUE TITLE
SSL issue with cloudflare images

### ISSUE THREAD

**Issue by HowardJeng** [JimmXinu/FanFicFare#560]
I tried using the calibre plugin to download a fic on archiveofourown.org (https://archiveofourown.org/works/26650795) but the image inside didn't get saved. The log showed an expired certificate error on the image, though when I checked the certificate it showed the expiration date was 2020-11-25.  Otherwise, the only odd thing I see about the certificate was that the image URL (https://cdn.discordapp.com/attachments/635168368055287811/759115490123055174/hp_treat.jpg) has a host of `cdn.discordapp.com` while the certificate was issued to `ssl711320.cloudflaressl.com`, which should be valid with the way cloudflare works. When I added `use_ssl_unverified_context:true` to my personal.ini and forced an overwrite, the image was saved properly.

AFAICT the same error occurred with https://archiveofourown.org/works/26646625 (Original download lacked image, though I didn't bother with the logs, just re-ran the download with `use_ssl_unverified_context:true` and the image was there.)

As a sanity check, I tried running the command line python script, but the log didn't seem to acknowledge the image existed at all, even if `--unverified_ssl` was used. Nor was the image in the resulting epub file with or without the switch. I'm not sure if this is a separate issue or if I'm just missing a logging switch.

Calibre log from failed download:
```Download 1 FanFiction Book(s)
FFF: INFO: 2020-09-28 03:21:56,375: calibre_plugins.fanficfare_plugin.jobs(53): FanFicFare v3.23.1
FFF: INFO: 2020-09-28 03:21:56,376: calibre_plugins.fanficfare_plugin.jobs(57): Adding jobs for URLs:
FFF: INFO: 2020-09-28 03:21:56,376: calibre_plugins.fanficfare_plugin.jobs(59): https://archiveofourown.org/works/26650795
FFF: INFO: 2020-09-28 03:22:12,602: calibre_plugins.fanficfare_plugin.jobs(95): Logfile for book ID 2223 (Swirly Glasses Squad)
FFF: INFO: 2020-09-28 03:22:12,603: calibre_plugins.fanficfare_plugin.jobs(96): url:(https://archiveofourown.org/works/26650795) id:(2223)

FFF: INFO: 2020-09-28 03:21:57,733: calibre_plugins.fanficfare_plugin.fanficfare.adapters.adapter_archiveofourownorg(164): url: https://archiveofourown.org/works/26650795/navigate?view_adult=true

FFF: INFO: 2020-09-28 03:21:57,733: calibre_plugins.fanficfare_plugin.fanficfare.adapters.adapter_archiveofourownorg(165): metaurl: https://archiveofourown.org/works/26650795?view_adult=true

FFF: DEBUG: 2020-09-28 03:21:57,734: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1149): #####################################

pagecache(GET) MISS: https://archiveofourown.org/works/26650795/navigate?view_adult=true

FFF: DEBUG: 2020-09-28 03:22:00,482: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1007): Encoding:utf8

FFF: DEBUG: 2020-09-28 03:22:00,483: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1149): #####################################

pagecache(GET) MISS: https://archiveofourown.org/works/26650795?view_adult=true

FFF: DEBUG: 2020-09-28 03:22:03,758: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1007): Encoding:utf8

FFF: DEBUG: 2020-09-28 03:22:03,926: calibre_plugins.fanficfare_plugin.fanficfare.adapters.adapter_archiveofourownorg(271): numChapters: (1)

FFF: INFO: 2020-09-28 03:22:04,050: calibre_plugins.fanficfare_plugin.jobs(246): write to C:\Users\hjeng\AppData\Local\Temp\calibre_mmbaawdg\fanficfare_ly12_a2b\bgmeta-jdzcs539.epub

FFF: INFO: 2020-09-28 03:22:04,051: calibre_plugins.fanficfare_plugin.fanficfare.writers.base_writer(182): Save directly to file: C:\Users\hjeng\AppData\Local\Temp\calibre_mmbaawdg\fanficfare_ly12_a2b\bgmeta-jdzcs539.epub

FFF: DEBUG: 2020-09-28 03:22:04,051: calibre_plugins.fanficfare_plugin.fanficfare.adapters.adapter_archiveofourownorg(421): Getting chapter text for: https://archiveofourown.org/works/26650795/chapters/64992877 index: 0

FFF: DEBUG: 2020-09-28 03:22:04,052: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1149): #####################################

pagecache(GET) MISS: https://archiveofourown.org/works/26650795/chapters/64992877?view_adult=true

FFF: DEBUG: 2020-09-28 03:22:09,973: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1007): Encoding:utf8

FFF: DEBUG: 2020-09-28 03:22:10,092: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1149): #####################################

pagecache(GET) MISS: https://cdn.discordapp.com/attachments/635168368055287811/759115490123055174/hp_treat.jpg

FFF: INFO: 2020-09-28 03:22:12,168: calibre_plugins.fanficfare_plugin.fanficfare.story(1267): Failed to load or convert image, 

parent:https://archiveofourown.org/works/26650795/chapters/64992877

skipping:https://cdn.discordapp.com/attachments/635168368055287811/759115490123055174/hp_treat.jpg

Exception: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired (_ssl.c:1123)>

FFF: INFO: 2020-09-28 03:22:12,178: calibre_plugins.fanficfare_plugin.fanficfare.writers.writer_epub(362): Saving EPUB Version 2.0


FFF: INFO: 2020-09-28 03:22:12,603: calibre_plugins.fanficfare_plugin.jobs(100): 
Download Results:
Add https://archiveofourown.org/works/26650795 Download epub completed, 1 chapters.

FFF: INFO: 2020-09-28 03:22:12,603: calibre_plugins.fanficfare_plugin.jobs(124): 
Add
https://archiveofourown.org/works/26650795
```
Log from command line execution:
```
$ fanficfare -d https://archiveofourown.org/works/26650795
FFF: DEBUG: 2020-09-28 19:29:00,583: cli.py(209):     OS Version:Linux-5.4.0-48-generic-x86_64-with-glibc2.29
FFF: DEBUG: 2020-09-28 19:29:00,583: cli.py(210): Python Version:3.8.2 (default, Jul 16 2020, 14:00:26) 
[GCC 9.3.0]
FFF: DEBUG: 2020-09-28 19:29:00,583: cli.py(211):    FFF Version:3.23.0
FFF: INFO: 2020-09-28 19:29:00,610: adapter_archiveofourownorg.py(164): url: https://archiveofourown.org/works/26650795/navigate
FFF: INFO: 2020-09-28 19:29:00,610: adapter_archiveofourownorg.py(165): metaurl: https://archiveofourown.org/works/26650795
FFF: DEBUG: 2020-09-28 19:29:00,610: configurable.py(1149): #####################################
pagecache(GET) MISS: https://archiveofourown.org/works/26650795/navigate
FFF: DEBUG: 2020-09-28 19:29:03,022: configurable.py(1007): Encoding:utf8
FFF: DEBUG: 2020-09-28 19:29:03,022: configurable.py(1149): #####################################
pagecache(GET) MISS: https://archiveofourown.org/works/26650795
FFF: DEBUG: 2020-09-28 19:29:05,528: configurable.py(1007): Encoding:utf8
FFF: DEBUG: 2020-09-28 19:29:05,758: adapter_archiveofourownorg.py(271): numChapters: (1)
FFF: INFO: 2020-09-28 19:29:05,884: base_writer.py(182): Save directly to file: Swirly Glasses Squad-ao3_26650795.epub
FFF: DEBUG: 2020-09-28 19:29:05,884: adapter_archiveofourownorg.py(421): Getting chapter text for: https://archiveofourown.org/works/26650795/chapters/64992877 index: 0
FFF: DEBUG: 2020-09-28 19:29:05,886: configurable.py(1149): #####################################
pagecache(GET) MISS: https://archiveofourown.org/works/26650795/chapters/64992877
FFF: DEBUG: 2020-09-28 19:29:08,478: configurable.py(1007): Encoding:utf8
FFF: INFO: 2020-09-28 19:29:08,616: writer_epub.py(362): Saving EPUB Version 2.0
FFF: DEBUG: 2020-09-28 19:29:08,713: cli.py(87): Successfully wrote 'Swirly Glasses Squad-ao3_26650795.epub'
```
```
$ fanficfare --unverified_ssl -d https://archiveofourown.org/works/26650795
FFF: WARNING: 2020-09-29 04:05:59,890: cli.py(200): Running with unverified SSL context(--unverified_ssl), security may be compromised.
FFF: DEBUG: 2020-09-29 04:05:59,898: cli.py(209):     OS Version:Linux-5.4.0-48-generic-x86_64-with-glibc2.29
FFF: DEBUG: 2020-09-29 04:05:59,898: cli.py(210): Python Version:3.8.2 (default, Jul 16 2020, 14:00:26) 
[GCC 9.3.0]
FFF: DEBUG: 2020-09-29 04:05:59,898: cli.py(211):    FFF Version:3.23.0
FFF: INFO: 2020-09-29 04:05:59,926: adapter_archiveofourownorg.py(164): url: https://archiveofourown.org/works/26650795/navigate
FFF: INFO: 2020-09-29 04:05:59,930: adapter_archiveofourownorg.py(165): metaurl: https://archiveofourown.org/works/26650795
FFF: DEBUG: 2020-09-29 04:05:59,937: configurable.py(1149): #####################################
pagecache(GET) MISS: https://archiveofourown.org/works/26650795/navigate
FFF: DEBUG: 2020-09-29 04:06:02,337: configurable.py(1007): Encoding:utf8
FFF: DEBUG: 2020-09-29 04:06:02,338: configurable.py(1149): #####################################
pagecache(GET) MISS: https://archiveofourown.org/works/26650795
FFF: DEBUG: 2020-09-29 04:06:04,826: configurable.py(1007): Encoding:utf8
FFF: DEBUG: 2020-09-29 04:06:05,011: adapter_archiveofourownorg.py(271): numChapters: (1)
FFF: INFO: 2020-09-29 04:06:05,141: base_writer.py(182): Save directly to file: Swirly Glasses Squad-ao3_26650795.epub
FFF: DEBUG: 2020-09-29 04:06:05,142: adapter_archiveofourownorg.py(421): Getting chapter text for: https://archiveofourown.org/works/26650795/chapters/64992877 index: 0
FFF: DEBUG: 2020-09-29 04:06:05,143: configurable.py(1149): #####################################
pagecache(GET) MISS: https://archiveofourown.org/works/26650795/chapters/64992877
FFF: DEBUG: 2020-09-29 04:06:07,786: configurable.py(1007): Encoding:utf8
FFF: INFO: 2020-09-29 04:06:07,934: writer_epub.py(362): Saving EPUB Version 2.0
FFF: DEBUG: 2020-09-29 04:06:08,032: cli.py(87): Successfully wrote 'Swirly Glasses Squad-ao3_26650795.epub'
```

**Comment by JimmXinu** [comment]
A few points:

1. Some stories on some sites have hundreds of images with dozens of broken image links.  FFF attempts to download images, but when they fail, FFF goes on relatively quietly.  This is by design.

1. Python's SSL depends on the cert chain either in python or in the operating system, depending on OS and configuration.  I don't have any control over what certs any particular user has to validate connections.  Browsers usually maintain their own cert store to be more up to date.   On Windows, the MS browsers will also update the OS cert store.  I've no idea how it works on your computer.

1. Python and cloudflare don't always get along.  Cloudflare is a large company who's mission includes preventing automatic tools from maliciously accessing sites under their protection.  Sometimes FFF for a particular user bumps into that and is blocked for a while.  Usually that user can later get through.  I'm honestly impressed Cloudflare can recognize and allow non-malicious tools access at all.

In this particular case, I do see the same error you do trying to fetch that image with a 5 line test program.  Which suggests this is more about point 3 than point 2, since I'm running on Windows.

As you've already found, `--unverified_ssl` and `use_ssl_unverified_context` are available, but they are officially 'unsupported'.  If they help you, great, but I make no promises and offer very little support for them.

**TL;DR:** There's nothing I can do about it, you've already found the only option I could offer; use it at your own risk.
